executor/src/witgen/jit/compiler.rs

```diff
@@ -0,0 +1,539 @@
+            lhs,
+            rhs,
+            expected_equal,
+        }) => format!(
+            "    assert!({} {} {});",
+            format_expression(lhs),
+            if *expected_equal { "==" } else { "!=" },
+            format_expression(rhs)
+        ),
+        Effect::MachineCall(id, arguments) => todo!(),
+    }
+}
+
+fn format_expression<T: FieldElement>(e: &SymbolicExpression<T, Cell>) -> String {
+    match e {
+        SymbolicExpression::Concrete(v) => format!("FieldElement::from({v})"),
+        SymbolicExpression::Symbol(cell, _) => cell_to_var_name(cell),
+        SymbolicExpression::BinaryOperation(left, op, right, _) => {
+            let left = format_expression(left);
+            let right = format_expression(right);
+            match op {
+                BinaryOperator::Add => format!("({left} + {right})"),
+                BinaryOperator::Sub => format!("({left} - {right})"),
+                BinaryOperator::Mul => format!("({left} * {right})"),
+                BinaryOperator::Div => format!("({left} / {right})"),
+                BinaryOperator::IntegerDiv => format!("integer_div({left}, {right})"),
+                BinaryOperator::BitAnd => format!("({left} & {right})"),
+                BinaryOperator::BitOr => format!("({left} | {right})"),
+            }
+        }
+        SymbolicExpression::UnaryOperation(op, inner, _) => {
+            let inner = format_expression(inner);
+            match op {
+                UnaryOperator::Neg => format!("-{inner}"),
+            }
+        }
+    }
+}
+
+fn cell_to_var_name(
+    Cell {
+        column_name,
+        row_offset,
+        ..
+    }: &Cell,
+) -> String {
+    // TODO this might still not be unique.
+    let column_name = column_name
+        .rfind("::")
+        .map_or(column_name.as_str(), |i| &column_name[i + 2..])
+        .replace("[", "_")
+        .replace("]", "_");
+    if *row_offset < 0 {
+        format!("{column_name}_u{}", -row_offset)
+    } else {
+        format!("{column_name}_d{row_offset}")
+    }
+}
+
+/// Returns the rust code containing utility functions given a first column id and a column count
+/// that is used to store the column table.
+fn util_code<T: FieldElement>(first_column_id: u64, column_count: usize) -> Result<String, String> {
+    if !(T::has_direct_repr() && (mem::size_of::<T>() == 8 || mem::size_of::<T>() == 4)) {
+        return Err(format!(
+            "Field {}not supported",
+            T::known_field()
+                .map(|f| format!("{f} "))
+                .unwrap_or_default()
+        ));
+    }
+
+    let int_type = if mem::size_of::<T>() == 8 {
+        "u64"
+    } else {
+        "u32"
+    };
+    let double_int_type = if mem::size_of::<T>() == 8 {
+        "u128"
+    } else {
+        "u64"
+    };
+    let modulus = T::modulus();
+
+    Ok(format!(
+        r#"#![allow(non_snake_case)]
+
+#[derive(Clone, Copy, Default)]
+#[repr(transparent)]
+struct FieldElement({int_type});
```

Comment:
Is there a plan to use optimized field implementations, like we have for Goldilocks?

Reply:
Yes, but I think we have to re-implement it here.